include/fast_gicp/gicp/lsq_registration.hpp

```diff
@@ -12,7 +12,86 @@ namespace fast_gicp {
 
 enum class LSQ_OPTIMIZER_TYPE { GaussNewton, LevenbergMarquardt };
 
-template<typename PointSource, typename PointTarget>
+template <int Dim>
+struct OptimizationParamProcessor {
+  virtual Eigen::Matrix<double, Dim, Dim> reduce_H(const Eigen::Matrix<double, 6, 6>& H_in) const = 0;
+  virtual Eigen::Matrix<double, Dim, 1> reduce_b(const Eigen::Matrix<double, 6, 1>& b_in) const = 0;
+  virtual Eigen::Matrix<double, 6, 1> expand_b(const Eigen::Matrix<double, Dim, 1>& b_in) const = 0;
+  using Ptr = std::shared_ptr<OptimizationParamProcessor<Dim>>;
+};
+
+template <>
+struct OptimizationParamProcessor<3> {
+  virtual Eigen::Matrix<double, 3, 3> reduce_H(const Eigen::Matrix<double, 6, 6>& H_in) const { return H_in.template topLeftCorner<3, 3>(); }
+  virtual Eigen::Matrix<double, 3, 1> reduce_b(const Eigen::Matrix<double, 6, 1>& b_in) const { return b_in.template topLeftCorner<3, 1>(); }
+  virtual Eigen::Matrix<double, 6, 1> expand_b(const Eigen::Matrix<double, 3, 1>& b_in) const {
+    Eigen::Matrix<double, 6, 1> d = Eigen::Matrix<double, 6, 1>::Zero();
+    d.topLeftCorner<3, 1>(3, 1) = b_in;
+    return d;
+  }
+  using Ptr = std::shared_ptr<OptimizationParamProcessor<3>>;
+};
+
+template <>
+struct OptimizationParamProcessor<6> {
+  virtual Eigen::Matrix<double, 6, 6> reduce_H(const Eigen::Matrix<double, 6, 6>& H_in) const { return H_in; }
+  virtual Eigen::Matrix<double, 6, 1> reduce_b(const Eigen::Matrix<double, 6, 1>& b_in) const { return b_in; }
+  virtual Eigen::Matrix<double, 6, 1> expand_b(const Eigen::Matrix<double, 6, 1>& b_in) const { return b_in; }
+  using Ptr = std::shared_ptr<OptimizationParamProcessor<6>>;
+};
+
+struct TranslationOnly : public OptimizationParamProcessor<3> {
+  Eigen::Matrix<double, 3, 3> reduce_H(const Eigen::Matrix<double, 6, 6>& H_in) const override { return H_in.template bottomRightCorner<3, 3>(); }
+  Eigen::Matrix<double, 3, 1> reduce_b(const Eigen::Matrix<double, 6, 1>& b_in) const override { return b_in.template bottomRightCorner<3, 1>(); }
+  Eigen::Matrix<double, 6, 1> expand_b(const Eigen::Matrix<double, 3, 1>& b_in) const override {
+    Eigen::Matrix<double, 6, 1> d = Eigen::Matrix<double, 6, 1>::Zero();
+    d.bottomRightCorner<3, 1>() = b_in;
+    return d;
+  }
+};
+
+template <bool B, bool... Args>
+struct count_true {
+  static constexpr int value = count_true<B>::value + count_true<Args...>::value;
+};
```

Comment:
i think these fold expressions need c++ 20, not sure how this can be realized with prior language standards.